audio/snd-util/adagio/ad.1
.TH AD 1 LOCAL
.SH NAME
ad \- play adagio or translate to midi
.SH SYNOPSIS
.B ad
[ options ] file [ options ]
.SH DESCRIPTION
.I Ad
plays an
.B Adagio
score on an fm sound card under
.I Linux
or translates it into the track portion of a standard
.B midi
file with output to
.IR stdout .
One or more such tracks can be combined into
form a standard
.B midi
file with the utility
.IR tracks .
.PP
If a file with the name supplied on the command line is not
found,
.I ad
supplies the suffix ".gio" and looks again.
.PP
The playing routine of
.I ad
is identical with that of the playing programs
.IR mp and xmp ,
and it also uses cfg configuration files in the same
way as those programs (except for track and channel
assignments), so please refer to mp(1) and
xmp(1) for details.
The Adagio scoring language does not currently provide
for setting chorus depth or reverberation.
.PP
There are the following command line options:
.TP
.BR \-external \ (or \ \-e )
means don't send any channels to the external midi port.
(Which channels will be external, if any, in the absence of
this flag, depends on a compilation option -- see the Makefile.)
.TP
.BR \-p
plays all parts except percussion with piano, ignoring program
requests.
.TP
.BR \-r \ chan[,prog]
records midi input from the external port in a new file "mpout.mid".
"Chan" is a number from 1-16, the channel for the midi commands in
the file, and "prog" is a number from 0-127, for the general midi
program to be used.  The output midi file contains a single track
which can be merged with the file being played by the separate
utility programs
.IR midt \ and \ tm .
.TP
.BR \-s
prevents treating any instruments as "solo" or non-polyphonic instruments.
A new note on a solo instrument cuts off any previous note (e.g.,
on a flute as opposed to piano).
.TP
.BR \-midi \ (or \ \-m )
translates instead of playing.
.TP
.B \-help
displays a usage summary.
.TP
.BR \-init \ (or \ \-i )
prevents a series of initialization commands for
each channel from being issued at the beginning of
the track.
.TP
.B \-v
gives some information about
which programs are used for which voices, and some
other things.
.TP
.B \-vv
gives even more information than
.B \-v
does.
.TP
.BR \-tune \ file
uses tuning information from the file.
.TP
.BR \-trace \ (or \ \-t )
issues tracing information.
.SH "SEE ALSO"
mp(1), setfm(1), tracks(1), midt(1), ``The CMU Midi Toolkit'' by Roger Dannenberg
.SH BUGS
This version of the Adagio language does not provide means to
set all midi controllers -- only numbers 1, 5, 64, and 65.
Consequently, other controller information is lost in the
translation from midi to adagio done by "mp -a".
.br
There is no provision for drum roll notes (notes > 81 on a
percussion channel).
.SH AUTHORS
Roger Dannenberg,
Roger.Dannenberg@SPICE.CS.CMU.EDU,
(Adagio and Transcribe interpreter code),
.br
Greg Lee, lee@uhunix.uhcc.hawaii.edu
(the rest).


audio/snd-util/adagio/ad.c
/*   Adagio compiler -- scan command line, call phase1, phase2 */

/*****************************************************************************
*	    Change Log
*  Date	    | Change
*-----------+-----------------------------------------------------------------
* 12-Mar-86 | Created
* 28-May-86 | Added calls to cmdline.c for command line parsing
*****************************************************************************/

#include <stdio.h>
#include "adagio.h"
#include "userio.h"
#include "cmdline.h"
#include "allphase.h"

event_type phase1();
void phase2();
void card_init(char *devname);

extern int verbose;
extern int piano_only;
extern int recording_track;
char score_na[255];  /* this will be available to phase1 and 2 */

/****************************************************************************
* Variables set by command line switches
****************************************************************************/

/****************************************************************************
* Data for command line parsing
****************************************************************************/
#define nswitches 15
static char *switches[nswitches] = 
    { "-help", "-v", "-vv",
      "-init", "-i",
      "-trace", "-t",
      "-external", "-e",
      "-midi", "-m",
      "-p", "-s",
      "-f", "-g" };

#define noptions 4
static char *options[noptions] = { "-tune", "-r", "-R", "-C" };
#define n_e_sw 2
static char *e_switches[n_e_sw] = { "-e", "-external" };

/****************************************************************************
*	Routines local to this module
****************************************************************************/
static void cmdline_help();

/****************************************************************************
*				 cmdline_help
* Effect: 
*	Prints out command line help
****************************************************************************/

static void cmdline_help()
{
    fprintf(stderr,"ad [options] filename [options]\n");
    fprintf(stderr,"	   -help	     this message\n");
    fprintf(stderr,"	   -init (-i)        initialize channels\n");
    fprintf(stderr,"	   -v    	     verbose\n");
    fprintf(stderr,"	   -vv   	     very verbose\n");
    fprintf(stderr,"	   -tune file	     use tuning from file\n");
    fprintf(stderr,"	   -trace (-t)       trace music\n");
    fprintf(stderr,"	   -midi (-m)        midi file on stdout\n");
    fprintf(stderr,"	   -p                piano only\n");
    fprintf(stderr, "	   -s                no solo instruments\n");
    fprintf(stderr, "	   -r  chan[,prog]   record new track\n");
    fprintf(stderr, "	   -e                no external synth\n");
    fprintf(stderr, "	   -g                no gus synth\n");
    fprintf(stderr, "	   -f                no fm synth\n");
}

/****************************************************************************
*				     main
* Inputs:
*	int argc: Count of arguments
*	char * argv[]: Vector of argument pointers
* Effect: 
*	Runs adagio
****************************************************************************/

void main(argc,argv)
   int argc;
   char * argv[];
{
    FILE *fp = NULL;
    char *s;	/* temp pointer to input file name */

    if (!cl_init(switches, nswitches, options, noptions, argv, argc))
	exit(1);

    score_na[0] = 0;

    if (cl_switch("-help")) {
	cmdline_help(); 
	return;
    }
    verbose = cl_switch("-v");
    really_verbose = cl_switch("-vv");
    if (really_verbose) verbose = really_verbose;
    piano_only = cl_switch("-p");
    exclude_fm = cl_switch("-f");
    exclude_gus = cl_switch("-g");
    no_solo = cl_switch("-s");
    extflag = (cl_nswitch(e_switches, n_e_sw) == NULL);
    if (cl_option("-r") != NULL) recording_track = true;

    /* get input file: */
    if ((s = cl_arg(1)) != NULL) strcpy(score_na, s);
    fp = fileopen(score_na, "gio", "r", "Name of Adagio score file");

    if (!cl_switch("-midi") && !cl_switch("-m"))
	card_init("/dev/sequencer");
    else {
	sb_dev = 0;
	card_info[0].nr_voices = 18;
    }

    phase2(phase1(fp));
}
